### PR TITLE
Migrate SCSS variables to CSS3 to enable auto dark and more features

### LINKED ISSUE #428: New color scheme with hybrid dark mode support
> Originally discussed in: https://github.com/kint-php/kint/issues/422#issuecomment-2381454524

In Kint 6, we can drop IE support and use [CSS variables](https://www.google.com/search?q=css+variables&ie=UTF-8&hl=en) together with media queries to detect dark mode support by the device or the browser and automatically apply a cool looking dark variant of that theme:

```css
@media (prefers-color-scheme: dark) {...}
```

This preferably needs to be applied to all themes currently residing in the resources directory.

While at it, the color scheme can be refreshed for a more modern and streamlined look, with more contrast.

### PR DESCRIPTION
**Pull request by DRSDavidSoft**
### Description
This is a preliminary PR to address #428. Changes so far include:

- Converted all SCSS variables to CSS variables instead where required (no scss variable remained after this change)

- Drop IE support (eliminating the need for base64 encoding of SVG images, reducing both file clutter and size)

- General cleanup to improve consistency across themes

**Note:** This PR currently _doesn't_ introduce visual changes. If there are any differences from the previous styles, please flag them. I’ve reviewed this but would appreciate a thorough check.

**Additional Requests:**
- I used `d($_SERVER);` to verify theme consistency. But it might not output use of all the Kint features, so could you please prepare a script for me that would make use of all the Kint features for a detailed comparison between the previous and old styles?

- For theme files, we currently have separate files for `solarized-light` and `solarized-dark`. Ideally, we should consolidate to a single `solarized.css` file with color variables adaptable to both themes. Users might prefer explicitly setting light or dark themes, though, and we should avoid disrupting existing user settings. I propose a “theme alias” feature in Kint, mapping old themes to a unified auto theme while applying `kint-dark` or `kint-light` classes as appropriate. This would allow us to ship just one file instead of three. Let me know if you agree, and I’ll work on a PR.

- Once decisions are finalized, another PR could be opened to refresh theme colors using professionally curated palettes in preparation for the Kint 6 release.

### Theme Variable Enhancements
- Added CSS variables for spacing, colors, and design elements for easier customization and theme management.

### Simplification and Consistency
- Replaced hardcoded values with CSS variables in order to enable declarative overriding (e.g. using `@media (prefers-color-scheme: dark)` query)
- Simplified complex selectors and properties to improve readability and maintainability.

The changes will help in maintaining a consistent design and make future modifications more straightforward.

### Tests
Verified the produced `css` in the compiled directory against the older versions by setting the following:

```php
RichRenderer::$theme = 'aante.css';
```

To ensure new styles exactly match previous themes.

### CODE REVIEW

**Comment by jnvsor** [comment]
Good work

> I propose adding a "theme alias" feature in Kint which is very simple, it would map the old theme files to the new auto theme, while also applying a kint-dark or kint-light class to the renderer.

Hmm I'd rather apply it to the body since there's only supposed to be "1 true stylesheet" for kint in a page. Are there any fancy frontend systems you know of that could break that? (ie. react)

I'll review this today

File: resources/sass/aante.scss
Comment (jnvsor) [inline resources/sass/aante.scss]: Let's add the media tag in a separate commit afterwards

File: resources/sass/aante.scss
Comment (jnvsor) [inline resources/sass/aante.scss:1]: Move `aante` back to `aante-light` for now (Makes diffing easier)

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: I'm concerned the variables might clobber user themes with common names like `--spacing`. We could namespace everything but I think a simpler solution would be to attach them to `.kint-rich` rather than `:root` since we control everything that should be in there anyway

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: Is there a technical reason for changing this? Are we heading towards a future without sass?

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss:52]: Could you split this out into its own commit?

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss:3]: I was worried this was without a unit for a reason but it looks like everywhere `$spacing` and `$color_size` were used used pixels

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: This is flipped

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: Could you move these up above `background`? (Also nice cleanup btw)

File: resources/sass/aante-dark.scss
Comment (jnvsor) [inline resources/sass/aante-dark.scss]: No technical reason for getting rid of the invert?

File: resources/sass/aante.scss
Comment (jnvsor) [inline resources/sass/aante.scss]: The original was `spacing * 6px + 1px` - there's a gap in the border (On firefox at least) without the extra pixel

File: resources/sass/original.scss
Comment (jnvsor) [inline resources/sass/original.scss]: Since these are specific to `original` I deliberately didn't bother putting them in variables. Maybe 1 and 2 should be since they're used more than once...

In that case what do you think of naming them `--gradient-start`/`--gradient-end`?

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: Broken. Missing `var()`

All of this stuff was carefully tweaked numbers while checking browsers but it looks like `calc` behaves differently and it'll need to be straightened out afterwards

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss]: Broken. Missing `var()`

**Review by jnvsor — CHANGES_REQUESTED** [review]
Additionally a weird rendering artifact only occurs on firefox not chrome:

image

I'll try to debug this myself

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: In my tests you can just set this to `var(--color-size)` and it will work now

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: Awesome, that's actually one point I forgot to mention. I was initially going to namespace the vars like `--kint-spacing`, but attaching them to `.kint-rich` makes sense too

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: The technical reason for the `--backdrop-color` change is that it needs to be overridable from other themes.

The reason behind changing `rgba(#fff, 0.9)` to `rgba(255, 255, 255, 0.9)` is because the former doesn't appear to be valid CSS3 color, while the latter is.  We could still use Sass-style colors, but it needs to be transpiled into regular CSS for it to work and I couldn't figure it out in the spur of the moment.

Are we heading towards a future without Sass? No, it doesn't make sense to stop using all the nice benefits of Sass, such as nested declarations. The intention is to only support using native CSS variables, so that they can be overridden in run time (rather than build time) by other themes, depending on the situation (e.g. using `media` queries), which can't be done using Sass-only variables.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss:52]: Sure, made this:

https://github.com/kint-php/kint/pull/430

I assumed it would be fine to integrate it here at first, in order to reduce clutter and avoid having to open multiple PRs.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss:3]: Right? It was also my initial assumption, however upon further look it appears that this was just done to make use of Sass calculation.

It makes sense to explicitly set the unit to `px` here rather than everywhere else in the CSS rewrite, correct? Also, a theme might use another unit of their choosing.

Using CSS' native `calc()` doesn't care _what_ the initial unit is, it just performs multiplication or division on it, which is our goal here.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: Good catch! Sorry for this, fixed in the newer commit that I will push soon.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: Thanks! Sure, done! Will push soon.

File: resources/sass/aante-dark.scss
Comment (DRSDavidSoft) [reply resources/sass/aante-dark.scss]: I wouldn't say a "technical" reason, but here are the reasons:
1. `invert(...)` is a Sass feature, not native to CSS. I could not figure out how to apply the colors so that it would transpile correctly.
2. In the upcoming commits, I plan to "refresh" the colors that is used. When @LoranRendel helped add `aante-dark` last year, they used the inverted colors, which is fine, but if we used "dark-optimized" values (that are not just the `invert()`-ed variant of the light ones), it can look better.

This means that if the chosen values are completely unrelated to those of the light variant, we can forgo using `invert()` altogether and just use custom dark-optimized values that look nicer.

File: resources/sass/aante.scss
Comment (DRSDavidSoft) [reply resources/sass/aante.scss]: Man, I am sorry I don't know why I missed these things, very important catch, thanks!

Fixed in the newer commit which I will push soon.

File: resources/sass/original.scss
Comment (DRSDavidSoft) [reply resources/sass/original.scss]: Exactly, I just named them `color1` and `color2` as a temporary placeholder.

I prefer to use `gradient-start` and `gradient-end`, just that since there are two gradients used here, maybe they can be named `primary-gradient-[start|end]` and `secondary-gradient-[start|end]`, I just didn't know which was which when I did it.

Alternatively, would be to make the entire gradient a variable instead. Then, instead of using CSS to apply them to the elements in question, the newly created `--background-image` variable can be used instead.

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: The `rgba(#fff, 0.9)` _is_ sass and it should compile to `rgba(255, 255, 255, 0.9)` automatically. I'm not sure why it won't, maybe sass explicitly disables it in css variables?

File: resources/sass/original.scss
Comment (jnvsor) [reply resources/sass/original.scss]: There are 2 gradients but `color3` `color4` and `highlight-color` are only used once so you might as well make them literals

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss:3]: I think the intent was to force the unit to be pixels (Since if you use something like em in a recursive structure you'll get ballooning sizes)

Also for line-height there's merit to unitless input (`line-height: 1.1` is roughly `line-height: calc(font-size * 1.1)`) but again we don't actually use it. I have no strong opinions one way or the other

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: Ugh. Got it working but the syntax sucks:

```scss
--backdrop-color: #{rgba(#FFF, 0.9)};
```

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: I must apologize again, I really don't know why I missed some of these 😅 

Anyways, piece of advice (~15 years of web dev) is that using this approach is not a sure thing across all the different screens, DPI scalings, fonts, OSes, etc.

Can you please include of how it is supposed to look? I believe we can include a SVG icon that would look way better, anyways.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: I believe using a `flex` solution is better than relying on `line-height`, that's an outdated method.

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: The problem is that we have 3 icons on the right:

* ⌕
* ↧
* ⇄

I don't know about your screen but on mine one of these is much smaller than the rest:

> image

It could just be my font but 1.25 has been pretty consistent for me

**Comment by DRSDavidSoft** [comment]
> Hmm I'd rather apply it to the body since there's only supposed to be "1 true stylesheet" for kint in a page. 

Good call, so setting the theme to an alias `aante-dark` will actually load the `aante` theme, and apply `kint-dark` class to the body itself

> Are there any fancy frontend systems you know of that could break that? (ie. react)
I don't think there would be a problem, although I don't think React would be used this way with PHP.

If a request from React is made to a PHP backend (e.g. Using Axiom), then it's most likely a "fetch" type request (à la ajax) and won't be directly rendered in the page.

There is one problem that comes to mind, which is that we will need to declare our styles as `.kint-dark .kint-rich { ... }` instead of `.kint-rich.kint-dark { ... }`. Actually, not a real problem, just something to have in mind.

**Comment by DRSDavidSoft** [comment]
Please review the requested changes and mark each conversation that is correctly resolved as such, if it wasn't resolved correctly please add a new reply.

Thanks! 🤝

**Comment by jnvsor** [comment]
> Good call, so setting the theme to an alias aante-dark will actually load the aante theme, and apply kint-dark class to the body itself

> If a request from React is made to a PHP backend (e.g. Using Axiom), then it's most likely a "fetch" type request (à la ajax) and won't be directly rendered in the page.

Yeah but you can't set a class on the body from kint in PHP, would have to be in JS. I can imagine different dumps with the light and dark version toggling the class on the body which is not what we want I presume.

Maybe it _should_ be a class like `.kint-rich.kint-dark` but that wouldn't work for the folder... I guess we could just have JS store the dark flag of first dump moved to the folder and use that for the folder class

File: resources/sass/aante.scss
Comment (DRSDavidSoft) [reply resources/sass/aante.scss]: Sure, I will keep it for now in this WIP PR, then I will move it to its own PR before merging this one.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: Excellent, and I agree the syntax is less than visually appealing, but where we _need_ to make use of Sass utils, it's acceptable.

I would say that after applying a "color refresh" to make it look better, some of these Sass definitions wouldn't be used as we will just apply an absolute CSS color.

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: Actually it could just be `line-height: 1` since we set the font-size to `--color-size` anyway (sorry for nitpick heh)

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss:3]: I agree, also there is the possibility of not using `line-height` for this and make use of other approaches.

Comment (jnvsor) [reply resources/sass/_base.scss:3]: The dom structure needs a rework anyway (It's too deep with too many elements and makes the browser take longer to render the dom than kint takes to render the markup) but that's a big change that touches everything including JS so it's definitely not on the table rn.

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: I am not applying any custom fonts or anything like that, but the magnifying glass on my screen clearly has the same width as other icons, while in your screenshot it does not:

image

I vote to use SVG icons here too. If you accept, I will whip a quick PR that uses Sass to load SVG icons directly into the CSS code.

One advantage this has is better theming support, like how the caret is themed.

Not lying, I disliked how these icons looked right from start when they were introduced, they don't "feel" like the rest of Kint, and I believe they can be improved if SVG icons are used instead.

Comment (jnvsor) [reply resources/sass/_base.scss]: Alrighty, whip me up an SVG then. I wrote the one in the caret by hand years ago.

I know you can make an svg spritesheet where each icon is on its own "layer" but I don't remember how to do it.

As far as the icons go other than the search icon (Which is pretty universally a magnifying glass) I have no particular preference. The ones in use were just easily usable as htmlentities.

**Comment by DRSDavidSoft** [comment]
> Maybe it _should_ be a class like `.kint-rich.kint-dark` but that wouldn't work for the folder... I guess we could just have JS store the dark flag of first dump moved to the folder and use that for the folder class

If I am not mistaken, the folder functionality already makes use of JS, right? If so, it makes sense for PHP to set the `kint-dark` class on the `kint-rich` element, then _when_ the folder feature is initialized, using JS we will check whether this class if set, and if so, copy it to the folder element.

> I can imagine different dumps with the light and dark version toggling the class on the body which is not what we want I presume.

I wouldn't say that it would make sense, but I believe _if_ the user can set the theme to be anything for each individual dump, it would make sense that they could also turn the dark mode on/off by specifying another "theme alias".

In conclusion, this makes more sense to me when I think about it, rather than setting `kint-dark` on the body itself.

> you can't set a class on the body from kint in PHP, would have to be in JS.

Quite right, the closest thing I can imagine is having a parent `.kint-dark` element, but none of this makes sense as opposed to just applying the `kint-dark` class directly on `kint-rich` element.

I guess _functionality speaking_, it doesn't make sense to have per-dump dark control, so a "global" dark mode control approach would've been nice. But _technically speaking_, setting it on the whole page doesn't make sense, and setting it on each individual dump makes more sense.

**Comment by jnvsor** [comment]
> If I am not mistaken, the folder functionality already makes use of JS, right? If so, it makes sense for PHP to set the kint-dark class on the kint-rich element, then when the folder feature is initialized, using JS we will check whether this class if set, and if so, copy it to the folder element.

Yeah that's exactly what I meant. Let's go with that (I can work on the php/js side) once this migration is done

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss:4]: This was originally intended for color swatches but ended up getting used for right side buttons too. It's basically the maximum height of blocks inside the bar. It should probably be renamed in a patch _after_ this one is live

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss:208]: I know with large dumps the reflows can get nasty. Could this have perf impacts?

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss]: For now I'll resolve this since it's equivalent to the old styles

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss]: Sure, I set it to `lint-height: normal` and also applied the flex to make sure it is properly centered, check the new commit to see

(Also don't worry, I have the exact same feelings regarding the nitpicks, even that 1px makes my OCD go creazy 😂 so it'll be awesome to make sure it looks fine without any pixel displacements, whatsoever)

Comment (jnvsor) [reply resources/sass/_base.scss]: > Depends on the user agent. Desktop browsers (including Firefox) use a default value of roughly 1.2, depending on the element's font-family

Per mdn: https://developer.mozilla.org/en-US/docs/Web/CSS/line-height#normal

That's a lot of conditions. `line-height: 1` is probably safer (Until your svg patch comes in anyway)

**Review by jnvsor — CHANGES_REQUESTED** [review]
Sorry I didn't press the big green button

**Comment by jnvsor** [comment]
@DRSDavidSoft any movement on this? 8.4 is now scheduled to release on the 21st so we have a little under 2 weeks.

I'd like to push an RC by the end of this weekend, but I still have more changes to make before then.

If you can't make it by then just take your time and we'll aim for 6.1

**Comment by DRSDavidSoft** [comment]
@jnvsor I was just a bit busy at work, if possible I would like to focus the PR only at migrating the variables for now to keep it simpler and allow us the time for it to be merged in `6.0`. Then, I will work on another PR to address the auto dark mode and the new icons. I hope that's also fine with you.

If so, can you please make a list of the changes that I need to apply for this PR to be merged? That would help me prioritize the changes as well as allowing #428 to be implemented in a later PR that will be based on this one.

**Comment by jnvsor** [comment]
Yeah that works for me

I'll just resolve everything and make a new review then

File: resources/sass/aante-dark.scss
Comment (jnvsor) [inline resources/sass/aante-dark.scss]: Put vars in .kint-rich (In all scss files)

In most children extending _base they already have a .kint-rich block so merge them rather than duplicating them

File: resources/sass/aante-light.scss
Comment (jnvsor) [inline resources/sass/aante-light.scss]: Remove this for now, we'll add it later

File: resources/sass/original.scss
Comment (jnvsor) [inline resources/sass/original.scss]: Change `--kint-color1` to `--gradient-start` and `kint-color2` to `--gradient-end`

Remove the other 3 vars. Since they're only used once just use literals

File: resources/sass/_base.scss
Comment (jnvsor) [reply resources/sass/_base.scss:208]: I'll perf check this myself

File: resources/sass/_base.scss
Comment (jnvsor) [inline resources/sass/_base.scss:397]: I'm not sure these changes won't break some layouts but I'll merge them first and test them myself later

**Review by jnvsor — CHANGES_REQUESTED** [review]

File: resources/sass/_base.scss
Comment (DRSDavidSoft) [reply resources/sass/_base.scss:397]: I don't see why the new approach would break anything, but we could also go this route as well:
`size: min(30px, 100%)`
which has the same effect.

Alternatively, if you found a breaking change with either of these CSS-only implementations, we can revert back to the SCSS solution.

File: resources/sass/original.scss
Comment (DRSDavidSoft) [reply resources/sass/original.scss]: Let's keep these as variables. My reasoning is that when we need to introduce the dark edition of this theme, we need the vars so that we could override them.

If we don't use variables here now, in the next PR those needs to be added later anyway.

Comment (jnvsor) [reply resources/sass/original.scss]: Good point... Then just leave them unnamed for now, I'll think of something to call them later

File: resources/sass/aante-dark.scss
Comment (DRSDavidSoft) [reply resources/sass/aante-dark.scss]: I applied this change, although I believe there could be some benefits for using the `:root` space, particularly if we use a short and concise prefix such as `--kt-`.

Would you prefer the following declarations? Otherwise, I'll put the variables in `.kint-rich` as requested.
```scss
:root {
  --kt-variable-type-color: ...
  --kt-variable-type-color-hover: ...
}
```

This would allow the users to easily override Kint themes using custom values later on.

File: resources/sass/original.scss
Comment (DRSDavidSoft) [reply resources/sass/original.scss]: The main purpose for using vars instead of literals is being able to override them in the _TBA_ dark theme variant of this theme.

File: resources/sass/aante-dark.scss
Comment (jnvsor) [reply resources/sass/aante-dark.scss]: Nah just leave it in `.kint-rich` - it's just as hard for a user to write

```css
.kint-rich {
  --variable-type-color: #F00;
}
```

As

```css
:root {
  --kt--variable-type-color: #F00;
}
```

**Comment by jnvsor** [comment]
I'm off to work now, I'll review tweak and merge later today

File: resources/sass/aante-dark.scss
Comment (DRSDavidSoft) [reply resources/sass/aante-dark.scss]: Actually what I mean is the ability to override the variable, but sure, I submitted the PR using the `.kint-rich` instead of `:root`

Note: using the `.kint-rich` approach, the user have to supply a more qualifying css selector in order to override their variables. e.g. add `kint-custom` class to the body, then insert overriding vars as `.kint-custom .kint-rich`, hopefully I was able to demonstrate my point.

Comment (jnvsor) [reply resources/sass/aante-dark.scss]: You can also do this with `:root .kint-rich` without touching the HTML, but even with that it's not a downgrade.

Before you had to style individual elements since you didn't have access to variables at all. Now you can still do it without more specific selectors if you make sure your css comes after kints (Which is admittedly pretty hard to do but easy enough for actual kint themes)

**Comment by jnvsor** [comment]
There's a few small things but I'll just take care of them here. For now could you hold off on making any more changes until I've pushed? Otherwise we're just going to cause conflicts

**Review by jnvsor — APPROVED** [review]

**Comment by jnvsor** [comment]
Merged in 63fcf330cd07bcd139230bb0354812d65260a029 with some tweaks after that.

The CSS could really use some love. I noticed the `.kint-parent` doesn't have a fixed height it's just flowing with the content. That could potentially be causing client-side perf issues.

Anyway I'll leave CSS here and push rc1 when I'm done with my changes

**Pull request closed without merging**
